File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/userUnsubscribe/UserUnsubscribeHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.userUnsubscribe

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.userUnsubscribe.UserUnsubscribeApi

import kangaroorewards.appsdk.core.io.EmptyResponse
import features.userUnsubscribe.serializeUserUnsubscribeState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class UserUnsubscribeHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/unsubscribe"

    override val eventChannel: String
        get() = "customer_sdk/events/unsubscribe"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return unsubscribe(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun unsubscribe(call: MethodCall): String? {
            val result = UserUnsubscribeApi().unsubscribe(
                campaignId = call.argument<String>("campaignId") as String,
                token = call.argument<String>("token") as String
            ).toJsonResult<EmptyResponse>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        UserUnsubscribeApi().userUnsubscribeState.serializeUserUnsubscribeState().observe { result ->
            pushSerializedResultToEventSink<UserUnsubscribeApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/userUnsubscribe/UserUnsubscribeHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class UserUnsubscribeHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/unsubscribe"

    var eventChannel: String = "customer_sdk/events/unsubscribe"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await UserUnsubscribeHandler.unsubscribe(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func unsubscribe(call: FlutterMethodCall) async -> String? {
        

        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
                        guard let campaignId = myArgs["campaignId"] as? String else {return nil}
                guard let token = myArgs["token"] as? String else {return nil}

         let result = try await UserUnsubscribeApi().unsubscribe(
                campaignId: campaignId,
                token: token
            ).serializeUserUnsubscribeApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        UserUnsubscribeApi().userUnsubscribeState
            .serializeUserUnsubscribeState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
